content/blog/2019-10-21-threads-not-as-libraries/index.md

```diff
@@ -0,0 +1,272 @@
+Then this program lacks data races, since reads and writes to `x` are logically synchronized.
+
+Further suppose that register promotion is determined to be a beneficial optimization, perhaps because the conditional is observed to rarely be taken.
+Since the Pthreads library calls are simply regarded as opaque function calls, the optimization will produce the following:
+
+```c
+r = x; 
+for (...) {
+  ...
+  if (mt) {
+    x = r; pthread_mutex_lock(...); r = x;
+  }
+  r = ... r...
+  if (mt) {
+    x = r; pthread_mutex_unlock(...); r = x;
+  }
+}
+x = r; 
+```
+
+In this case, register promotion introduces unsynchronized reads and writes to `x`, creating a data race.
+
+
+## Allowing Data Races
+In the face of these issues, it’s clear that, indeed, threads cannot be implemented as a library. Instead, the thread specification must be formally incorporated into the language specification, to ensure that compilers do not break the former.
+
+The task of formally specifying Pthreads presents an opportunity for rethinking its synchronization-oriented paradigm.
+The author identifies that synchronization is not always desirable, and that it requires a precise interleaving of expensive atomic operations:
+>The cost of atomic operations and memory barriers varies widely, but is often comparable to that of a hundred or more register-to-register instructions, even in the absence of a cache miss. For example, on some Pentium 4 processors, hardware instructions to atomically update a memory location require well over 100 processor cycles, and these can also double as one of the cheaper mechanisms for ensuring that a store operation becomes visible to other threads before a subsequent load. 
+
+Instead of synchronization, the author argues in favor of a paradigm that allows data races and relies on atomic operations by showing performance gains from such a paradigm shift. 
+>As a result of the high cost of these hardware instructions, and the even higher cost of the pthread primitives built on them, there are a small number of cases in which synchronization performance is critical, and more careful and direct use of the hardware primitives, together with less constrained use of shared variables, is essential. In some cases it may also be necessary to avoid deadlock issues inherent in lock-based programming[7], or desirable because a different parallel programming model is preferable for an application (cf. [32]).
+
+To demonstrate this claim, the author compares the performance of an implementation of the Sieve of Eratosthenes algorithm and a tracing garbage collector implemented under the synchronization and data raceful paradigm.
+
+Four implementations are evaluated when run with 1, 2, and 4 threads.  
+Two of these use synchronization: one using mutexes, and another using spinlocks.  
+These are compared to two implementations with data races: one that makes use of atomic memory operations, and one that unsafely uses ordinary memory operations.
```

Comment:
I suppose this is the crux of the matter: do we say that the atomic-using versions have data races? From a modern (2019) perspective, we would say "no," because the C++11 standard gave synchronization semantics to atomics. But from a 2004 perspective, perhaps. In any case, it is clear that the version with atomics is morally "better" than the ordinary-memory-op one.

Reply:
See above: I think the author needed to make a distinction between race conditions and data races.